Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp

```diff
@@ -365,16 +371,62 @@ void Animations::resume()
         animation.resume();
 }
 
-void Animations::apply(Animation::ApplicationResult& applicationResults, MonotonicTime time)
+void Animations::apply(Animation::ApplicationResult& applicationResults, MonotonicTime time, Animation::KeepInternalState keepInternalState)
 {
-    for (auto& animation : m_animations)
-        animation.apply(applicationResults, time);
-}
+    Vector<Animation*> translateAnimations;
+    Vector<Animation*> rotateAnimations;
+    Vector<Animation*> scaleAnimations;
+    Vector<Animation*> transformAnimations;
+    Vector<Animation*> leafAnimations;
 
-void Animations::applyKeepingInternalState(Animation::ApplicationResult& applicationResults, MonotonicTime time)
-{
-    for (auto& animation : m_animations)
-        animation.applyKeepingInternalState(applicationResults, time);
+    for (auto& animation : m_animations) {
+        switch (animation.keyframes().property()) {
+        case AnimatedProperty::Translate:
+            translateAnimations.append(&animation);
+            break;
+        case AnimatedProperty::Rotate:
+            rotateAnimations.append(&animation);
+            break;
+        case AnimatedProperty::Scale:
+            scaleAnimations.append(&animation);
+            break;
+        case AnimatedProperty::Transform:
+            transformAnimations.append(&animation);
+            break;
+        default:
+            leafAnimations.append(&animation);
+        }
+    }
+
+    if (!translateAnimations.isEmpty() || !rotateAnimations.isEmpty() || !scaleAnimations.isEmpty() || !transformAnimations.isEmpty()) {
+        applicationResults.transform = TransformationMatrix();
```

Comment:
Now you know applicationResults.transform is no longer std::nullopt, how about adding:
```
auto& transform = applicationResults.transform.value();
```
and avoid the `applicationResults.transform->multiply(...)` calls below and simly use `transform.multiply(...)` - wdyt?